Support JSON as source for parametrized tests arguments (#492)
junit-pioneer/junit-pioneer: docs/json.adoc

:page-title: JSON
:page-description: Extends JUnit Jupiter with `@JsonFileSource`, a parametrized test that creates test based on a JSON Source

> The `@JsonFileSource` annotation lets you use JSON files from the classpath or local file system to create parametrized tests.

Reviewer: I propose to slightly change the doc's structure. Add another sentence or two here what to look out for (single vs multiple arguments; nested arrays) and then add `== Method Arguments` and make the next two headers on single and multiple argument methods h3.

Author: Will doo that

By default the root of the source will be treated as candidate for the test arguments.
If the root is an object then the entire object will be one argument, if the root is an array then every element of the array will be one argument.

There is also `JsonFileSource#data` that can be used to tell the extraction mechanism to use the element with that name to look for the source of the data.

> Depending on how the test method parameters the extraction of the values might differ.

Reviewer: ```suggestion
Depending on the test method parameters, the extraction of the values might differ.
```

== Single Argument Methods

> When the method has a single argument then the json object argument will be converted to that type.

Reviewer: ```suggestion
If the method has a single argument, the json object argument will be converted to that type.
```

.Argument type
[source,java]
----
public class Customer {
	public String name;
	public String height;

	@Override
    public String toString() {
        return "Customer{" + "name='" + name + '\'' + ", height=" + height + '}';
    }
}
----

.JSON Source
[source,json]
----
[
  {
    "name": "Luke",
    "height": 172
  },
  {
    "name": "Yoda",
    "height": 66
  }
]
----

[source,java]
----
@ParametrizedTest
@JsonFileSource(resources = "/customers.json")
void singleCustomer(Customer customer) {
    assertThat(Collections.singleton(tuple(customer.getName(), customer.getHeight())))
            .containsAnyOf(
                tuple("Luke", 172),
                tuple("Yoda", 66)

>             );

Reviewer: I appreciate the attention to detail, but I don't think we need to flesh out the test. 😉

Author: 😄  I'll remove them from the docs

}
----

This parametrized test will generate the following tests:

* [1] Customer{name='Luke', height=172}
* [2] Customer{name='Yoda', height=66}

It is also possible to extract only a single element from each argument object by using the `@Param` annotation.

[source,java]
----
@ParametrizedTest
@JsonFileSource(resources = "/customers.json")
void singleCustomer(@Param("name") String customerName) {
    assertThat(customername)
            .isIn("Luke", "Yoda");
}
----

This parametrized test will generate the following tests:

* [1] Luke
* [2] Yoda

=== Multiple Argument Methods

> When the method has multiple arguments then each json object argument will be deconstructed to each of the method arguments.

Reviewer: ```suggestion
If the method has multiple arguments, each json object argument will be deconstructed to each of the method arguments.
```

By default, the method argument name will be used for locating the element that needs to be taken from the json object.
You can also use `@Param` to give the name of the element that needs to be extracted.

[IMPORTANT]
====
If your test sources are not compiled using the `--parameters` flag then the names of the arguments will not be like they are written in the source code.
In that the situation you need to use `@Param` instead.
====

Using the same `customers.json` and the following test

[source,java]
----
@ParametrizedTest
@JsonFileSource(resources = "/customers.json")
void deconstructFromArray(@Param("name") String customerName, @Param("height") int height) {
    assertThat(Collections.singleton(tuple(customerName, height)))
					.containsAnyOf(
							tuple("Luke", 172),
							tuple("Yoda", 66)

> 					);

Reviewer: As above.

}
----

This parametrized test will generate the following tests:

* [1] Luke, 172
* [2] Yoda, 66

> === Extracting nested array

Reviewer: ```suggestion
== Extracting nested array
```

Shouldn't this be h2 - it applies to single-argument methods as well, right?

Sometimes we want to extract a nested array instead of the root element.
For this purpose `JsonFileSource#data` can be used.

.Customer with nested array
[source,json]
----
{
  "name": "Luke",
  "height": 172,
  "vehicles": [
    {
      "name": "Snowspeeder",
      "length": 4.5
    },
    {
      "name": "Imperial Speeder Bike",
      "length": 3
    }
  ]
}
----

Here we want to test the customer vehicles.
The test for this will look like:

[source,java]
----
@ParameterizedTest
@JsonFileSource(resources = "customer-luke.json", data = "vehicles")
void lukeVehicles(@Param("name") String name, @Param("length") double length) {
    assertThat(Collections.singleton(tuple(name, length)))
            .containsAnyOf(
                tuple("Snowspeeder", 4.5),
                tuple("Imperial Speeder Bike", 3d)

>             );

Reviewer: As above.

}
----

This parametrized test will generate the following tests:

* [1] Snowspeeder, 4.5

> * [2] Imperial Speeder Bike, 3

Reviewer: Can you add a section on thread safety like the other extensions have? Will probably look [like this](https://junit-pioneer.org/docs/range-sources/#thread-safety).

Author: Makes sense. I'll do that
